### PR TITLE
add support for volume plugins

### PR DESCRIPTION
**Pull request by wlan0**
@ibuildthecloud

### CODE REVIEW

File: cattle/plugins/docker/storage.py [hunk 1]
```diff
@@ -201,6 +248,14 @@ def _do_volume_remove(self, volume, storage_pool, progress):
             if container is None:
                 return
             remove_container(docker_client(), container)
+        elif self._is_managed_volume(volume):
+            try:
+                v = DockerConfig.storage_api_version()
+                docker_client(version=v).remove_volume(volume.name)
+            except APIError as e:
```
Comment (ibuildthecloud) [inline cattle/plugins/docker/storage.py]: don't catch then raise

File: cattle/plugins/docker/storage.py [hunk 2]
```diff
@@ -194,6 +227,20 @@ def _is_volume_removed(self, volume, storage_pool):
 
             return not os.path.exists(path)
 
+    def _is_managed_volume(self, volume):
+        try:
```
Comment (ibuildthecloud) [inline cattle/plugins/docker/storage.py]: check driver not None and name is set and driverOpts field exists

File: cattle/plugins/docker/storage.py [hunk 3]
```diff
@@ -174,8 +174,34 @@ def _get_volume_storage_pool_map_data(self, obj):
         }
 
     def _is_volume_active(self, volume, storage_pool):
+        if not self._is_managed_volume(volume):
+            return True
+        if volume.data['fields']['driver'] is None:
```
Comment (ibuildthecloud) [inline cattle/plugins/docker/storage.py]: move to _is_managed_volume

File: cattle/plugins/docker/storage.py [hunk 4]
```diff
@@ -186,6 +212,13 @@ def _is_volume_removed(self, volume, storage_pool):
             return container is None
         else:
             path = self._path_to_volume(volume)
+            try:
```
Comment (ibuildthecloud) [inline cattle/plugins/docker/storage.py]: add elseif for _is_managed_volume()

File: cattle/plugins/docker/storage.py [hunk 5]
```diff
@@ -174,8 +174,34 @@ def _get_volume_storage_pool_map_data(self, obj):
...
+            return True
+        try:
+            v = DockerConfig.storage_api_version()
+            docker_client(version=v).inspect_volume(volume.name)
+        except APIError:
+            return False
         return True
 
+    def _do_volume_activate(self, volume, storage_pool, progress):
+        driver = ""
```
Comment (ibuildthecloud) [inline cattle/plugins/docker/storage.py]: check for is_managed and then assume fields are there

**Comment by cjellick** [comment]
closing in favor of #188

**Pull request closed without merging**